bundle_install_process.go

```diff
@@ -84,26 +91,61 @@ func (ip BundleInstallProcess) ShouldRun(layer packit.Layer, workingDir string)
 		}
 	}
 
-	cachedSHA, ok := layer.Metadata["cache_sha"].(string)
+	cachedSHA, ok := metadata["cache_sha"].(string)
 	cacheMatch := ok && cachedSHA == sum
 	shouldRun := !cacheMatch || !rubyVersionMatch
 
 	return shouldRun, sum, rubyVersion, nil
 }
 
+// Execute will configure and install a set of gems into a layer location using
+// the Bundler CLI.
+//
+// First, to configure the Bundler environment, Execute will copy the local
+// Bundler configuration, if any, into the target layer path. The configuration
+// file created in the layer will become the defacto configuration file by
+// setting `BUNDLE_USER_CONFIG` in the local environment while executing the
+// subsequent Bundle CLI commands. The configuration will then be modifed with
+// any settings specific to the invocation of Execute.  These configurations
+// will override any settings previously applied in the local Bundle
+// configuration.
+//
+// Once fully configured, Execute will run "bundle install" as a child process.
+// During the execution of the "bundle install" process, Execute will have
+// configured the command to use any locally vendored cache, enabling offline
+// execution.
 func (ip BundleInstallProcess) Execute(workingDir, layerPath string, config map[string]string) error {
 	localConfigPath := filepath.Join(workingDir, ".bundle", "config")
+	backupConfigPath := filepath.Join(workingDir, ".bundle", "config.bak")
 	globalConfigPath := filepath.Join(layerPath, "config")
+
+	err := os.RemoveAll(globalConfigPath)
+	if err != nil {
+		return err
+	}
+
 	if _, err := os.Stat(localConfigPath); err == nil {
 		err := os.MkdirAll(layerPath, os.ModePerm)
 		if err != nil {
 			return err
 		}
 
+		if _, err := os.Stat(backupConfigPath); err == nil {
+			err = fs.Copy(backupConfigPath, localConfigPath)
```

Comment:
I'm not sure if I fully understand for the `backupConfigPath`. Could you explain the need a bit more?

Reply:
This is because of a strange quirk in how Bundler works. It writes down some of the global configuration options into the local configuration file when you run `bundle install`. One of these options is the `without` option.

When the `gems` build plan entry is required for both `build` and `launch`, the buildpack will end up running the `bundle install` process 2 times, once for `build` gems and once for `launch` gems. In each case we are copying the local configuration into the layer and then modifying it with extra configuration. I found that in this case, the first invocation of `bundle install` would modify the local configuration file. Then, on the second pass, we would carry those modifications over into the `launch` layer, which is not really what we want. We want to start again with a clean slate where the local configuration file only contains those settings that were originally configured by the user.

So, I need to make a backup of the local configuration before `bundle install` is run so that I can record what that configuration contained when the user provided it to the buildpack. Then, on subsequent invocations of the `bundle install` process, the backup is restored to replace any modifications to the local configuration that might have taken place.

Reply:
That's a helpful explanation, thank you